### ISSUE TITLE
pkgs/desktops/gnome/extensions/update-extensions.py is broken

### ISSUE THREAD

**Issue by rhoriguchi** [NixOS/nixpkgs#190483]
### Describe the bug
For some strange reason there are a lot of duplicates when I try to update. To fix this, I tried to filter them out. But when I do that, it filters out some extensions which are not duplicated so somehow the `pk` is not unique.

```patch
diff --git a/pkgs/desktops/gnome/extensions/update-extensions.py b/pkgs/desktops/gnome/extensions/update-extensions.py
index 9117ea27040..aa9b623cee8 100755
--- a/pkgs/desktops/gnome/extensions/update-extensions.py
+++ b/pkgs/desktops/gnome/extensions/update-extensions.py
@@ -245,10 +245,14 @@ def scrape_extensions_index() -> List[Dict[str, Any]]:
             else:
                 raise
 
-    # `pk` is the primary key in the extensions.gnome.org database. Sorting on it will give us a stable,
-    # deterministic ordering.
-    extensions.sort(key=itemgetter("pk"))
-    return extensions
+    # `pk` is the primary key in the extensions.gnome.org database.
+
+    # Remove duplicate entries.
+    used = set()
+    extensions = [extension for extension in extensions if extension["pk"] not in used and (used.add(extension["pk"]) or True)]
+
+    # Sorting on it will give us a stable, deterministic ordering.
+    return sorted(extensions, key=itemgetter("pk"))
 
 
 if __name__ == "__main__":
```

### Steps To Reproduce
1. `./pkgs/desktops/gnome/extensions/update-extensions.py`

### Notify maintainers
- @piegamesde
- @jtojnar

---

As side note, `arcmenu@arcmenu.com` version `17` has an invalid zip.

```patch
diff --git a/pkgs/desktops/gnome/extensions/update-extensions.py b/pkgs/desktops/gnome/extensions/update-extensions.py
index 9117ea27040..50641288aff 100755
--- a/pkgs/desktops/gnome/extensions/update-extensions.py
+++ b/pkgs/desktops/gnome/extensions/update-extensions.py
@@ -98,25 +98,29 @@ def generate_extension_versions(
     # Download information once for all extension versions chosen above
     extension_info_cache: Dict[ExtensionVersion, Tuple[str, str]] = {}
     for extension_version in sorted(set(extension_versions.values())):
-        logging.debug(
-            f"[{uuid}] Downloading v{extension_version}"
-        )
-        extension_info_cache[extension_version] = \
-            fetch_extension_data(uuid, str(extension_version))
+        # TODO remove when arcmenu@arcmenu.com version != 17, this version has a broken zip
+        if uuid != "arcmenu@arcmenu.com" and extension_version != "17":
+            logging.debug(
+                f"[{uuid}] Downloading v{extension_version}"
+            )
+            extension_info_cache[extension_version] = fetch_extension_data(uuid, str(extension_version))
 
     # Fill map
     extension_versions_full: Dict[ShellVersion, Dict[str, str]] = {}
     for shell_version, extension_version in extension_versions.items():
-        sha256, metadata = extension_info_cache[extension_version]
-
-        extension_versions_full[shell_version] = {
-            "version": str(extension_version),
-            "sha256": sha256,
-            # The downloads are impure, their metadata.json may change at any time.
-            # Thus, we back it up / pin it to remain deterministic
-            # Upstream issue: https://gitlab.gnome.org/Infrastructure/extensions-web/-/issues/137
-            "metadata": metadata,
-        }
+        # TODO remove when arcmenu@arcmenu.com version != 17, this version has a broken zip
+        if uuid != "arcmenu@arcmenu.com" and extension_version != "17":
+            sha256, metadata = extension_info_cache[extension_version]
+
+            extension_versions_full[shell_version] = {
+                "version": str(extension_version),
+                "sha256": sha256,
+                # The downloads are impure, their metadata.json may change at any time.
+                # Thus, we back it up / pin it to remain deterministic
+                # Upstream issue: https://gitlab.gnome.org/Infrastructure/extensions-web/-/issues/137
+                "metadata": metadata,
+            }
+
     return extension_versions_full
 
 
```

**Comment by piegamesde** [comment]
That's a weird one. I can confirm that the scrape results in duplicate entries. I tried exporting the raw data with

```python
    with open(updater_dir_path / "extensions-debug.json", "w") as out:
        json.dump(extensions, out, ensure_ascii=False, indent=2)
```

and then analyzing it with [`jq`](https://jqplay.org/). To see the duplicates, see `[.[] | "\(.uuid), \(.pk)"] | sort`, leave out the sort at the end and you'll notice that the duplicates are not necessarily adjacent (which was my first assumption). I then inspected one of the duplicated extensions with `[.[] | select(.uuid == "empathy-gwibber-in-usermenu-3.6")]` and convinced myself that the duplicates are indeed identical.

However, I did not manage to find any cases where the `pk` is used by different extension UUIDs.
